### PR TITLE
Fix Nmap Crash when host is a list of dicts, instead of a dict

### PR DESCRIPTION
**Pull request by ohachimOs**
In Nmap `scan_results`, there is a `host` field that can either be a `List` or a `Dict`, in `_emit_network_scan_finding`, we had it working for `Dict` only. This PR makes it so that if host is a list, it will loop over the host list and report a vuln for each of them.

image

### CODE REVIEW

**Review by nmasdoufi-ol — COMMENTED** [review]
Please add a description for the PR.

File: agent/nmap_agent.py [hunk 1]
```diff
@@ -253,15 +253,29 @@ def _emit_network_scan_finding(
...
-                    for domain_dict in domains:
-                        domain = domain_dict.get("@name", "")
+            hosts = scan_results.get("nmaprun", {}).get("host", {})
+            if isinstance(hosts, dict):
+                hosts = [hosts]
+            for host in hosts:
+                domains = host.get("hostnames", {})
+                ports = host.get("ports", {}).get("port", "")
+                address = host.get("address", {})
+                if domains is not None and len(domains.values()):
```
Comment (nmasdoufi-ol) [inline agent/nmap_agent.py:263]: len(domains.values()) > 0 or what?

File: agent/nmap_agent.py [hunk 2]
```diff
@@ -253,15 +253,29 @@ def _emit_network_scan_finding(
...
-            domains = host.get("hostnames", {})
-            ports = host.get("ports", {}).get("port", "")
-            address = host.get("address", {})
-            if domains:
-                domains = domains.get("hostname", {})
-                if isinstance(domains, List):
-                    for domain_dict in domains:
-                        domain = domain_dict.get("@name", "")
+            hosts = scan_results.get("nmaprun", {}).get("host", {})
+            if isinstance(hosts, dict):
```
Comment (ybadaoui-ostorlab) [inline agent/nmap_agent.py:257]: explicit here

File: tests/nmap_agent_test.py [hunk 3]
```diff
@@ -1214,3 +1316,22 @@ def testAgentNmapOptions_whenServiceHasNoProduct_reportsFingerprintzzz(
...
+    """Test host in scan results is a list instead of a dict, the agent should not
+    crash, instead it completes its emitting process for each of the hosts in the host list."""
+    mocker.patch(
+        "agent.nmap_wrapper.NmapWrapper.scan_domain",
+        return_value=(SCAN_RESULT_HOST_AS_LIST, HUMAN_OUTPUT),
+    )
+
+    nmap_test_agent.process(domain_msg)
+
+    assert len(agent_mock) == 16
```
Comment (mohsinenar) [inline tests/nmap_agent_test.py:1337]: Can you please add explicit checks on the messages.

**Review by mohsinenar — DISMISSED** [review]

File: tests/nmap_agent_test.py (hunk 3 above)
Comment (adnaneserrar) [reply tests/nmap_agent_test.py:1337]: +1

File: agent/nmap_agent.py [hunk 4]
```diff
@@ -253,15 +253,29 @@ def _emit_network_scan_finding(
...
-                    for domain_dict in domains:
-                        domain = domain_dict.get("@name", "")
+            hosts = scan_results.get("nmaprun", {}).get("host", {})
+            if isinstance(hosts, dict) is True:
+                hosts = [hosts]
+            for host in hosts:
+                domains = host.get("hostnames", {})
+                ports = host.get("ports", {}).get("port", "")
+                address = host.get("address", {})
+                if domains is not None and len(domains.values()) > 0:
```
Comment (oussamaessaji) [inline agent/nmap_agent.py:263]: Reverse the check to decrease the indentations

File: tests/nmap_agent_test.py (hunk 3 above)
Comment (3asm) [reply tests/nmap_agent_test.py:1337]: Yes, please assert on the content.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/nmap_agent.py [hunk 5]
```diff
@@ -253,15 +253,29 @@ def _emit_network_scan_finding(
...
+                    domains = domains.get("hostname", {})
+                    if isinstance(domains, List) is True:
```
Comment (deadly-panda) [inline agent/nmap_agent.py:265]:
```suggestion
                    if isinstance(domains,list) is True:
```

File: agent/nmap_agent.py (hunk 4 above)
Comment (deadly-panda) [reply agent/nmap_agent.py:263]: > Reverse the check to decrease the indentations

Yes, & the whole thing can go into a function..

Comment (ohachimOs) [reply agent/nmap_agent.py:263]: I can't reverse this condition, because it is chained with other `elif`s, inside a loop, can't use `continue`

**Review by nmasdoufi-ol — APPROVED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**